### PR TITLE
Searching for hidden objects

### PR DESCRIPTION
**Pull request by Owcaman**
A code bit that enables you to search through a directory and find folders - either visible or hidden.

### CODE REVIEW

File: stand-alone/com/microstrategy/samples/hiddenObjects.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
+/* FOREWORD:
```
Comment (tiagosiebler) [inline stand-alone/com/microstrategy/samples/hiddenObjects.java]: Following the pattern here so far, can you move this java class to `stand-alone/com/microstrategy/samples/searching/hiddenObjects.java`?

File: stand-alone/com/microstrategy/samples/hiddenObjects.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+3. The GUID of the folder that you want to search through - the sample one used is the Public Objects folder.
+
+Apart from those details the code is ready to run.
+
+Made with love by Edgar, based on [278906/DE124645/KB16830]
+Did not use object.populate() but search.getResults() instead. Chose to search folders as needed it for practical purposes.
+
+Find the code below, and ENJOY. */
+
+package com.edgar.standalone;
```
Comment (tiagosiebler) [inline stand-alone/com/microstrategy/samples/hiddenObjects.java]: Once you've moved this class, make sure to edit this package line to `package com.microstrategy.samples.searching;`

File: stand-alone/com/microstrategy/samples/hiddenObjects.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+import com.microstrategy.web.objects.WebObjectSource;
+import com.microstrategy.web.objects.WebObjectsException;
+import com.microstrategy.web.objects.WebObjectsFactory;
+import com.microstrategy.web.objects.WebSearch;
+import com.microstrategy.webapi.EnumDSSXMLApplicationType;
+import com.microstrategy.webapi.EnumDSSXMLObjectFlags;
+import com.microstrategy.webapi.EnumDSSXMLObjectTypes;
+import com.microstrategy.webapi.EnumDSSXMLSearchFlags;
+
+public class FirstClass {
```
Comment (tiagosiebler) [inline stand-alone/com/microstrategy/samples/hiddenObjects.java]: Rename `FirstClass` to the name of this Java file (`public class hiddenObjects {`)

File: stand-alone/com/microstrategy/samples/hiddenObjects.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+System.out.println("Initiating the session.");
+
+serverSession = factory.getIServerSession();
+serverSession.setServerName("SERVER"); 
+serverSession.setServerPort(0);
+serverSession.setProjectName("MicroStrategy Tutorial");
+serverSession.setLogin("Administrator"); 
+serverSession.setPassword(""); 
+serverSession.setApplicationType(EnumDSSXMLApplicationType.DssXmlApplicationCustomApp);
```
Comment (tiagosiebler) [inline stand-alone/com/microstrategy/samples/hiddenObjects.java]: Can you re-use the session logic from `SessionManager.getSessionWithDetails()`? It's less repeat code this way, see here for a usage example:
https://github.com/MicroStrategy/web-sdk-samples/blob/master/stand-alone/com/microstrategy/samples/beans/reportbean/ExportPDF.java#L34

File: stand-alone/com/microstrategy/samples/hiddenObjects.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+search.types().add(EnumDSSXMLObjectTypes.DssXmlTypeFolder);
+search.submit();	
+
+System.out.println("Search submitted.");
+
+WebFolder folderWithResults = search.getResults();
+System.out.println("Filtering the results.");
+
+int size = 0;
+size = folderWithResults.size();
```
Comment (tiagosiebler) [inline stand-alone/com/microstrategy/samples/hiddenObjects.java]: You can combine these lines:
```
int size = folderWithResults.size();
```

File: stand-alone/com/microstrategy/samples/hiddenObjects.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+System.out.println("All of the folders from within the search directory: ");
+
+for (java.util.Enumeration e = folderWithResults.elements(); e.hasMoreElements();) {
+    System.out.println(e.nextElement());
+   }
```
Comment (tiagosiebler) [inline stand-alone/com/microstrategy/samples/hiddenObjects.java]: Slight cleanup, can you try this instead?
```
// get list of results into Enumeration
Enumeration<WebObjectInfo> results = folderWithResults.elements();

// iterate through each result, while results remain
while (results.hasMoreElements()) {
  WebObjectInfo result = results.nextElement();
  System.out.println(result.getName());
}
```

http://www.java2s.com/Code/JavaAPI/java.util/EnumerationhasMoreElements.htm

File: stand-alone/com/microstrategy/samples/hiddenObjects.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+System.out.println("Size of the folder is: " + size);
+
+	}
```
Comment (tiagosiebler) [inline stand-alone/com/microstrategy/samples/hiddenObjects.java]: Indentation is a bit quirky. Once you've finished these edits, select all (ctrl + a) then from the "Source" menu, choose "Correct Indentation":
Screenshot 2019-07-01 at 09 54 32

File: stand-alone/com/microstrategy/samples/hiddenObjects.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+serverSession = factory.getIServerSession();
+serverSession.setServerName("SERVER"); 
+serverSession.setServerPort(0);
+serverSession.setProjectName("MicroStrategy Tutorial");
+serverSession.setLogin("Administrator"); 
+serverSession.setPassword(""); 
+serverSession.setApplicationType(EnumDSSXMLApplicationType.DssXmlApplicationCustomApp);
+
+System.out.println("Performing the search.");
```
Comment (Dholmblad) [inline stand-alone/com/microstrategy/samples/hiddenObjects.java]: @Owcaman Minor suggestion, I would extract out the two actions you are performing here into separate methods and invoke them from the main method. Or extract them into reusable classes. cc @tiagosiebler

File: stand-alone/com/microstrategy/samples/searching/hiddenObjects.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+		System.out.println("All of the folders from within the search directory: ");
+
+		// Getting a list of results into Enumeration
+		Enumeration<WebObjectInfo> results = folderWithResults.elements();
+
+		// Iterating through each result, while results remain
+		while (results.hasMoreElements()) {
+			WebObjectInfo result = results.nextElement();
+			System.out.println(result.getName());
+			}
```
Comment (tiagosiebler) [inline stand-alone/com/microstrategy/samples/searching/hiddenObjects.java]: As discussed, you have an extra brace here at line 87 - try to run this sample in eclipse.

File: stand-alone/com/microstrategy/samples/hiddenObjects.java (hunk 8 above)
Comment (Owcaman) [reply stand-alone/com/microstrategy/samples/hiddenObjects.java]: Done!

File: stand-alone/com/microstrategy/samples/searching/hiddenObjects.java (hunk 9 above)
Comment (Owcaman) [reply stand-alone/com/microstrategy/samples/searching/hiddenObjects.java]: Fixed it. Thanks!

**Pull request merged**